qpdf-convert-server.py

```diff
@@ -0,0 +1,204 @@
+    print(msg, end=suffix, flush=True, file=sys.stderr)
+
+def die(msg):
+    '''Qrexec wrapper for displaying error messages'''
+    logging.error(msg)
+    sys.exit(1)
+
+def send(data):
+    '''Qrexec wrapper for sending text data to the client's STDOUT'''
+    print(data, flush=True)
+
+def send_b(data):
+    '''Qrexec wrapper for sending binary data to the client's STDOUT'''
+    sys.stdout.buffer.write(data)
+    sys.stdout.buffer.flush()
+
+def recv_b(size=None):
+    '''Qrexec wrapper for receiving data from a client'''
+    untrusted_data = sys.stdin.buffer.read(size)
+    return untrusted_data
+
+def recvline_b():
+    '''Qrexec wrapper for receiving a line of data from a client'''
+    untrusted_data = sys.stdin.buffer.readline()
+    return untrusted_data
+
+
+###############################
+#        Image-related
+###############################
+
+def send_img_dimensions(png_path):
+    cmd_width = ['identify', '-format', '%w', png_path]
+    cmd_height = ['identify', '-format', '%h', png_path]
+
+    try:
+        untrusted_width = subprocess.run(cmd_width, capture_output=True,
+                                         check=True).stdout.decode()
+        untrusted_height = subprocess.run(cmd_height, capture_output=True,
+                                         check=True).stdout.decode()
+    except subprocess.CalledProcessError:
+        die("Failed to gather dimensions... Aborting")
+
+    send(f'{untrusted_width} {untrusted_height}')
+
+def send_rgb_file(rgb_path):
+    with open(rgb_path, 'rb') as f:
+        data = f.read()
+        send_b(data)
+
+def pdf_to_png(pagenum, pdf_path, png_path):
+    png_filename = os.path.splitext(png_path)[0]
+    cmd = ['pdftocairo', pdf_path, '-png', '-f', str(pagenum), '-l',
+           str(pagenum), '-singlefile', png_filename]
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        cmd = ['convert', pdf_path, f'png:{png_path}']
+        try:
+            subprocess.run(cmd, check=True)
+        except subprocess.CalledProcessError:
+            die(f'Page {pagenum} conversion failed (PDF->PNG): {err}')
+
+def png_to_rgb(pagenum, png_path, rgb_path):
+    depth = 8
+    cmd = ['convert', png_path, '-depth', str(depth), f'rgb:{rgb_path}']
+
+    try:
+        subprocess.run(cmd, check=True)
+    except subprocess.CalledProcessError:
+        die(f'Page {pagenum} conversion failed (PNG->RGB): {err}')
+
+
+###############################
+#         File-related
+###############################
+
+def create_tmp_files():
+    '''Create temporary file for storing page images and the untrusted PDF'''
+    Files = namedtuple('Files', ['pdf', 'png', 'rgb'])
+    suffixes = ('', '.png', '.rgb')
+    paths = []
+
+    for suffix in suffixes:
+        with NamedTemporaryFile(prefix='qpdf-conversion-', suffix=suffix) as f:
+            paths.append(f.name)
+
+    for path in paths:
+        with open(path, 'wb') as f:
+            f.write(b'')
```

Comment:
This is wrong. `NamedTemporaryFile()` specifically create the file in a way that avoid collision. Once it's removed, you can't just reuse the name without extra checks to preserve this property. What you can do, is to add `delete=False` attribute to `NamedTemporaryFile()`. Nicer solution would be to put the code that actually use those files inside `with NamedTemporaryFile` context manager, but since you loop over file suffixes that would be more refactor.

Reply:
Fixed in local repo.

Reply:
> `NamedTemporaryFile()` specifically create the file in a way that avoid collision. Once it's removed, you can't just reuse the name without extra checks to preserve this property.

Alright, so the image file creation code is a bit different from before but you should get the idea. My initial fix a while back was to have the page number be a part of the temporary filename so that we're guaranteed no collisions.

```python
def create_rep(page, initial, final):
    with NamedTemporaryFile(prefix=f"qpdf-conversion-{page}-") as f:
        rep = Representation(initial=Path(f"{f.name}.{initial}"), final=Path(f"{f.name}.{final}"))
        rep.initial.touch()
        rep.final.touch()

    return rep
```

But then I learned you can open 2 files with `with open` and about the `delete=False` parameter, I think this is my best option so far:

```python
def create_rep(page, initial, final):
    with NamedTemporaryFile(prefix="qpdf-conversion-", suffix=f".{initial}", delete=False) as f_initial,
         NamedTemporaryFile(prefix="qpdf-conversion-", suffix=f".{final}", delete=False) as f_final:
        rep = Representation(initial=Path(f_initial.name), final=Path(f_final.name)

    return rep
```


> Nicer solution would be to put the code that actually use those files inside `with NamedTemporaryFile` context manager, but since you loop over file suffixes that would be more refactor.

I agree, but I don't see how it could work with how I've used asyncio. You can see that I create the temporary files for each page in the first loop. But I simply start the render tasks in that loop, saving my awaits for the second loop when all the renders have started. I don't see where I would put the `with open`. It obviously can't be in the second loop as that's already too late. But it also can't be in the first loop since once a render task is appended to `tasks`, the temporary files will close.

```python
async def run(pdf_path, pagenums, max_tasks=0):
    results = []
    tasks = []
    reps = []
    limit = max_tasks if max_tasks > 0 else pagenums

    for page in range(1, limit + 1):
         rep = create_rep(page, "png", "rgb")  # create image files
         reps.append(rep)
         tasks.append(asyncio.create_task(render(page, pdf_path, rep)))

    for task in tasks:
        try:
            results.append(await task)
        except subprocess.CalledProcessError:
            await cancel_tasks(tasks)
            delete_reps(reps)
            raise
...

Reply:
Another option would be to move `with NamedTemporaryFile` into `render()` function. Still some would require `delete=False` (the output rgb one), but not all of them anymore.

Or maybe do something different: use `TemporaryDirectory` for the whole processing (wrap the whole `run()` code with it) and generate filenames simply based on page numbers? This would also ease cleanup - simply remove the directory and don't worry about forgetting some files.

Reply:
> Or maybe do something different: use `TemporaryDirectory` for the whole processing (wrap the whole `run()` code with it) and generate filenames simply based on page numbers? This would also ease cleanup - simply remove the directory and don't worry about forgetting some files.

AHHHH this is such a good idea. Will implement.

Reply:
Fixed in ae014aa